src/network/core/game_info.cpp

```diff
@@ -264,27 +274,35 @@ void DeserializeNetworkGameInfo(Packet *p, NetworkGameInfo *info)
 	/* Update the documentation in game_info.h on changes
 	 * to the NetworkGameInfo wire-protocol! */
 
+	GameInfoNewGRFMode newgrf_mode = GAME_INFO_NEWGRF_MODE_SHORT;
 	switch (game_info_version) {
 		case 5:
 			info->join_key = p->Recv_string(NETWORK_JOIN_KEY_LENGTH);
+			newgrf_mode = (GameInfoNewGRFMode)p->Recv_uint8();
 			FALLTHROUGH;
 
 		case 4: {
-			GRFConfig **dst = &info->grfconfig;
-			uint i;
-			uint num_grfs = p->Recv_uint8();
-
-			/* Broken/bad data. It cannot have that many NewGRFs. */
-			if (num_grfs > NETWORK_MAX_GRF_COUNT) return;
-
-			for (i = 0; i < num_grfs; i++) {
-				GRFConfig *c = new GRFConfig();
-				DeserializeGRFIdentifier(p, &c->ident);
-				HandleIncomingNetworkGameInfoGRFConfig(c);
-
-				/* Append GRFConfig to the list */
-				*dst = c;
-				dst = &c->next;
+			if (newgrf_mode != GAME_INFO_NEWGRF_MODE_NONE) {
+				GRFConfig **dst = &info->grfconfig;
+				uint i;
+				uint num_grfs = p->Recv_uint8();
+
+				/* Broken/bad data. It cannot have that many NewGRFs. */
+				if (num_grfs > NETWORK_MAX_GRF_COUNT) return;
+
+				for (i = 0; i < num_grfs; i++) {
+					GRFConfig *c = new GRFConfig();
+					DeserializeGRFIdentifier(p, &c->ident);
+					std::string name = {};
+					if (newgrf_mode == GAME_INFO_NEWGRF_MODE_FULL) {
```

Comment:
Do you know the average number of times the same NewGRF is used by different servers? Like NewGRF A is used by 10 servers, NewGRF B by 5, NewGRF C by 1, so the average is about 16/5 ~= 5. That could be used to influence choosing an even more efficient way.
You would then be creating a list of used NewGRFs at the top of the packet with the whole listing, and here you pass indices into that list. Those indices can probably be uint16 but stick with uint32 for now for an added safety buffer. At an average of 2 you would be saving about 16 + NewGRF name length bytes per NewGRF, with an average at 3 you would save about 32 + 2 * NewGRF name length, and so forth.

Reply:
There are tons of more efficient ways honestly. I liked the suggestion to apply compression on such packets .. that also sound really useful, and maybe means we don't even need a trick like yours. But that needs more work and benchmarking.

For now, I went for a solution that was simple and elegant, rather than complicated and touching a lot of stuff. Future work could optimize this by a lot. But I rather not do that in this PR :) If you agree, ofc.

Btw, this last commit is purely here as otherwise after this PR we would have lost functionality we had before the PR.